### PR TITLE
pcp: Align API with draft-patton-cfrg-vdaf-00

### PR DESCRIPTION
**Pull request by cjpatton**
Replace `Value` with trait `Type`, implementations of which provide the
FLP functionality described in the spec. In particular, each `Type`
implementation specifies how measurements are encoded and validated. The
FLP system is now implemented as a set of derived methods on a generic
`Type`. However, the type itself does not subsume the input that is
validated.
    
Shifting the API boundary this way makes it easier to integrate the FLP
into `prio3`.

### CODE REVIEW

File: src/pcp.rs
Comment (tgeoghegan) [inline src/pcp.rs:287]: I'm kind of surprised we never use `self.proof_len()` in this function. Should we at least check that the `proof_len` computed here is the expected value?

File: src/pcp.rs
Comment (tgeoghegan) [inline src/pcp.rs:291]:
```suggestion
    /// Query an input and proof and return the verifier message, encoded as [`Self::verifier_len`] field elements.
```

File: src/pcp.rs
Comment (tgeoghegan) [inline src/pcp.rs:321]: Would it be simpler to do `if self.gadget.len() == query_rand.len() {}` once, before iterating `self.gadget()`? Then I think you could also drop the check for `query_rand.len() > shim.len()` on line 364, since `shim.len()` should always be the same as `self.gadget.len()`.

 Additionally I think you could zip `self.gadget().into_iter()` with `query_rand.into_iter()` and avoid indexing into `query_rand`, and then you don't need to `.enumerate()`, but either way is fine with me.

File: src/pcp.rs
Comment (tgeoghegan) [inline src/pcp.rs:415]: In the unlikely event that `poly_eval` returns a value longer than `shim[idx].arity`, we will end up with more than `data_len` elements in `verifier`. Could we check that `verifier.len() == self.verifier_len()` and error out if so?

File: src/pcp.rs
Comment (tgeoghegan) [inline src/pcp.rs:422]: Could check that `verifier.len() == self.verifier_len()`

File: src/pcp/types.rs
Comment (tgeoghegan) [inline src/pcp/types.rs:248]: Any reason not to do this now, and make `Histogram::new` fallible? Same goes for `Sum::new`.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: src/pcp/types.rs
Comment (cjpatton) [inline src/pcp/types.rs:128]: I meant to resolve these "XXX"s before starting review. I'll resolve them before we merge.

File: src/pcp.rs
Comment (cjpatton) [reply src/pcp.rs:415]: Hmm, I'm not sure I understand the problem. Are you concerned about an overflow? `poly_eval` always returns exactly one field element. In any case, if a `.push()` would exceed the limits of the memory allocated for the vector, the vector's memory will be extended.

File: src/pcp.rs
Comment (cjpatton) [reply src/pcp.rs:422]: This is done implicitly. Before parsing the next chunk from `verifier`, we check if `verifier` has enough memory (and fail if not). For the same reason I gave for `proof`, I think it's if we compute what `self.verifier_len()` is supposed to return to make it easier to catch incorrect implementations of this call.

Here again we could check that the computed verifier length matches `self.verifier_len()`.

File: src/pcp/types.rs
Comment (cjpatton) [reply src/pcp/types.rs:248]: The reason to do this in the constructor is that we only need to validate that `self.buckets` is well-formed once. As it stands, we end up re-validating this every time we call `encode`.

File: src/pcp.rs
Comment (tgeoghegan) [reply src/pcp.rs:155]: Yeah, I think that's correct in all cases.

File: src/pcp.rs
Comment (tgeoghegan) [reply src/pcp.rs:291]: I agree with your observation that this is programmer error so a panic feels appropriate.

File: src/pcp.rs
Comment (tgeoghegan) [reply src/pcp.rs]: You're right. I think it might have if you didn't have `let _ =` (which I like, because it explicitly tells the reader "this returns a value and we're intentionally ignoring it"). Might be worth filing an issue on clippy.

File: src/pcp.rs
Comment (tgeoghegan) [reply src/pcp.rs:412]: Oh, good point. My thinking is that this module should defend itself against errors in implementation of a `Type`, but you're right, `poly_eval` can't be implemented wrong, so we have a reasonable guarantee that the vector will contain the right number of elements. It might still be nice to check that the `Type`'s `verifier_len()` implementation is correct and panic -- your call.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]
I think we still need the checks on `Type::{proof_len, verifier_len}` and maybe `input_len` as discussed and then this is good to go.

File: src/pcp.rs
Comment (cjpatton) [reply src/pcp.rs:315]: Done.

File: src/pcp.rs
Comment (cjpatton) [reply src/pcp.rs:445]: I'll add an assertion, like we do for proof length in `prove`.

File: src/pcp.rs
Comment (cjpatton) [reply src/pcp.rs:453]: Done.

**Review by tgeoghegan — APPROVED** [review]

**Comment by cjpatton** [comment]
Squashed.

**Pull request merged**
